mindmeld/models/embedder_models.py

```diff
@@ -0,0 +1,184 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2015 Cisco Systems, Inc. and others.  All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains the embedder model class.
+"""
+from abc import ABC, abstractmethod
+import pickle
+import logging
+import os
+import string
+import numpy as np
+
+from .. import path
+from .helpers import register_embedder
+from .taggers.embeddings import WordSequenceEmbedding
+
+
+logger = logging.getLogger(__name__)
+
+try:
+    from sentence_transformers import SentenceTransformer
+
+    register_bert = True
+except ImportError:
+    logger.warning("Must install the extra [bert] to use the built in embbedder.")
+    register_bert = False
+
+
+class Embedder(ABC):
+    """
+    Base class for embedder model
+    """
+
+    def __init__(self, app_path, **kwargs):
+        """Initializes an embedder.
+        """
+        self.model_name = kwargs.get("model_name", "default")
+        self.cache_path = path.get_embedder_cache_file_path(
+            app_path, kwargs.get("embedder_type", "default"), self.model_name
+        )
+
+        folder = os.path.dirname(self.cache_path)
+        if not os.path.isdir(folder):
```

Comment:
Will there ever be a case that multiple `Embedder` instances will initalize concurrently in different threads or processes? If so, you'll want some additional safety around file operations here--there's a potential race condition here where two threads could simultaneously fail the check for the directory, then both try to create it. The same idea applies to other file system operations throughout this class.

If there will only ever be one instance, this will work fine the way it is.

Reply:
In the current code, there will only ever be one instance at a time (see `create_embedder_model` function). I can still add some additional safety if it's cleaner that way - do you have any docs/pointers on that?

Reply:
I believe @tmehlinger 's scenario happens when two MindMeld application pointing to the same app_path are running happen to reach this code path simultaneously. I don't think this scenario is very likely. Moreover, there aren't any adverse effects in clearing the cache apart from longer processing time. @tmehlinger Let us know if these reasons are sufficient to address the issue.